File: luogu/P1198-stack+binarysearch.cpp
/*
Solution can be found at P1198.py
*/

#include <iostream>
using namespace std;
const int N = 200010;
long long stacks[N];
int idxs[N], size = 0;
int binary_search(int num) {
    int low = 0, high = size - 1;
    while (true) {
        if (low >= high) {
            if (idxs[low] >= num) {
                return low;
            } else {
                return low + 1;
            }
        }
        int mid = (low + high) / 2;
        if (idxs[mid] < num) {
            low = mid + 1;
        } else if (idxs[mid] > num) {
            high = mid - 1;
        } else {
            return mid;
        }
    }
}
int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int M;
    long long D;
    scanf("%d %lld\n", &M, &D);

    int idx_now = 0;
    long long t = 0ll;
    for (int i = 0; i < M; i++) {
        char oper;
        long long n;
        scanf("%c %lld\n", &oper, &n);
        if (oper == 'A') {
            
            long long val = (t + n) % D;
            while (true) {
                if (size == 0) { break; }
                if (stacks[size - 1] > val) { break; }
                size--;
            }
            stacks[size] = val;
            idxs[size] = idx_now;
            size++;
            idx_now++;
        } else if (oper == 'Q') {
            int idx_use = binary_search(idx_now - int(n));
            printf("%lld\n", stacks[idx_use]);
            t = stacks[idx_use];
        } else {
            continue;
        }
    }


    return 0;
}

File: luogu/P1198-stack+binarysearch.py
###
# Solution: monotonic stack + binary search, but TLE with python
# we can hold a monotonic decreasing stack to store the max value from each position to final.
# then we just need to binary search for the given index len(A) - L, the first no less than this index in the stack
# store the largest value from len(A)-L to the end of A
###

def binary_search(stack, num):
    low, high = 0, len(stack) - 1
    while True:
        if low >= high:
            if stack[low][1] >= num:
                return low
            else:
                return low + 1

        mid = (low + high) >> 1
        if stack[mid][1] < num:
            low = mid + 1
        elif stack[mid][1] > num:
            high = mid - 1
        else:
            return mid

M, D = input().split()
M, D = int(M), int(D)

stack = []
idx_now = 0
t = 0
for _ in range(M):
    oper, n = input().split()
    n = int(n)
    if oper == 'A':
        val = (n + t) % D
        while True:
            if len(stack) == 0:
                break
            if stack[-1][0] > val:
                break
            stack.pop()
        stack.append([val, idx_now])
        idx_now += 1
    else:
        idx_use = binary_search(stack, idx_now-n)
        print(stack[idx_use][0])
        t = stack[idx_use][0]